telemetry/jetbrains/src/main/kotlin/software/aws/toolkits/telemetry/generator/GenerationFunctions.kt

```diff
@@ -0,0 +1,170 @@
+
+    val function = functionBuilder.build()
+    addFunction(function)
+    // Result is special cased to generate a function that accepts true/false instead of a Result
+    if (metric.metadata?.any { it.type == RESULT } != true) {
+        return
+    }
+    val resultFunction = FunSpec.builder(functionName)
+    val resultParameters = parameters.map {
+        if (it.name == RESULT) {
+            ParameterSpec.builder(SUCCESS, BOOLEAN).build()
+        } else {
+            it
+        }
+    }
+    resultFunction
+        .addParameters(resultParameters)
+        .generateResultOverloadFunctionBody(function, resultParameters)
+        .addKdoc(metric.description)
+    addFunction(resultFunction.build())
+}
+
+fun buildParameters(metric: Metric, types: List<TelemetryMetricType>): List<ParameterSpec> {
+    val projectParameter = ClassName("com.intellij.openapi.project", "Project").copy(nullable = true)
+    val list = mutableListOf<ParameterSpec>()
+    list.add(ParameterSpec.builder("project", projectParameter).defaultValue("null").build())
+    list.addAll(metric.metadata?.map { metadata -> metadata.metadataToParameter(types) } ?: listOf())
+    list.add(ParameterSpec.builder("value", DOUBLE).defaultValue("1.0").build())
+    list.add(ParameterSpec.builder("createTime", Instant::class).defaultValue("Instant.now()").build())
+    return list
+}
+
+fun Metadata.metadataToParameter(types: List<TelemetryMetricType>): ParameterSpec {
+    val telemetryMetricType = types.find { it.name == type } ?: throw IllegalStateException("Type $type not found in types!")
+
+    val typeName = if (telemetryMetricType.allowedValues != null) {
+        ClassName(PACKAGE_NAME, telemetryMetricType.name.toTypeFormat())
+    } else {
+        telemetryMetricType.type?.getTypeFromType() ?: com.squareup.kotlinpoet.STRING
+    }.copy(nullable = required == false)
+
+    val parameterSpec = ParameterSpec.builder(telemetryMetricType.name.toArgumentFormat(), typeName)
+    if (required == false) {
+        parameterSpec.defaultValue("null")
+    }
+    return parameterSpec.build()
+}
+
+fun FunSpec.Builder.generateFunctionBody(metric: Metric): FunSpec.Builder {
+    val telemetryClient = MemberName("software.aws.toolkits.jetbrains.services.telemetry", "TelemetryService")
+    val metricUnit = MemberName("software.amazon.awssdk.services.toolkittelemetry.model", "Unit")
+    addStatement("%M.getInstance().record(project) { ", telemetryClient)
+    addStatement("datum(%S) {", metric.name)
+    addStatement("createTime(createTime)")
+    addStatement("unit(%M.${(metric.unit ?: MetricUnit.NONE).name})", metricUnit)
+    addStatement("value(value)")
+    metric.metadata?.forEach {
+        generateMetadataStatement(it, "${it.type.toArgumentFormat()}.toString()")
+    }
+    addStatement("}}")
+
+    return this
+}
+
+fun FunSpec.Builder.generateResultOverloadFunctionBody(originalFunction: FunSpec, parameters: List<ParameterSpec>): FunSpec.Builder {
```

Comment:
private?

Reply:
set a lot of these private + made them actual builders